pubsublite/internal/wire/acks.go

```diff
@@ -0,0 +1,241 @@
+
+import (
+	"container/list"
+	"fmt"
+	"sync"
+)
+
+// AckConsumer is the interface exported from this package for acking messages.
+type AckConsumer interface {
+	Ack()
+}
+
+// ackedFunc is invoked when a message has been acked by the user.
+type ackedFunc func(*ackConsumer)
+
+// ackConsumer is used for handling message acks. It is attached to a Message
+// and also stored within the ackTracker until the message has been acked by the
+// user.
+type ackConsumer struct {
+	// The message offset.
+	Offset int64
+	// Bytes released to the flow controller once the message has been acked.
+	MsgBytes int64
+
+	// Guards access to fields below.
+	mu    sync.Mutex
+	acked bool
+	onAck ackedFunc
+}
+
+func newAckConsumer(offset, msgBytes int64, onAck ackedFunc) *ackConsumer {
+	return &ackConsumer{Offset: offset, MsgBytes: msgBytes, onAck: onAck}
+}
+
+func (ac *ackConsumer) Ack() {
+	ac.mu.Lock()
+	defer ac.mu.Unlock()
+
+	if ac.acked {
+		return
+	}
+	ac.acked = true
+	if ac.onAck != nil {
+		ac.onAck(ac)
+	}
+}
+
+func (ac *ackConsumer) IsAcked() bool {
+	ac.mu.Lock()
+	defer ac.mu.Unlock()
+	return ac.acked
+}
+
+// Clear onAck when the ack can no longer be processed. The user's ack would be
+// ignored.
+func (ac *ackConsumer) Clear() {
+	ac.mu.Lock()
+	defer ac.mu.Unlock()
+	ac.onAck = nil
+}
+
+// Represents an uninitialized cursor offset. A sentinel value is used instead
+// if an optional to simplify cursor comparisons (i.e. -1 works without the need
+// to check for nil and then convert to int64).
+const nilCursorOffset int64 = -1
+
+// ackTracker manages outstanding message acks, i.e. messages that have been
+// delivered to the user, but not yet acked. It is used by the committer and
+// wireSubscriber, so requires its own mutex.
+type ackTracker struct {
+	// Guards access to fields below.
+	mu sync.Mutex
+	// All offsets before and including this prefix have been acked by the user.
+	ackedPrefixOffset int64
+	// Outstanding message acks, strictly ordered by increasing message offsets.
+	outstandingAcks *list.List // Value = *ackConsumer
+}
+
+func newAckTracker() *ackTracker {
+	return &ackTracker{
+		ackedPrefixOffset: nilCursorOffset,
+		outstandingAcks:   list.New(),
+	}
+}
+
+// Push adds an outstanding ack to the tracker.
+func (at *ackTracker) Push(ack *ackConsumer) error {
+	at.mu.Lock()
+	defer at.mu.Unlock()
+
+	// These errors should not occur unless there is a bug in the client library
+	// as message ordering should have been validated by subscriberOffsetTracker.
+	if ack.Offset <= at.ackedPrefixOffset {
+		return errOutOfOrderMessages
+	}
+	if elem := at.outstandingAcks.Back(); elem != nil {
+		lastOutstandingAck, _ := elem.Value.(*ackConsumer)
+		if ack.Offset <= lastOutstandingAck.Offset {
+			return errOutOfOrderMessages
+		}
+	}
+
+	at.outstandingAcks.PushBack(ack)
+	return nil
+}
+
+// Pop processes outstanding acks and updates `ackedPrefixOffset` until an
+// unacked message is found.
+func (at *ackTracker) Pop() {
```

Comment:
Pop() is called [within the subscriber](https://github.com/tmdiep/google-cloud-go/blob/6f7a727a628b79d580060aaa2ecdbc7ad723dae0/pubsublite/internal/wire/subscriber.go#L254) and CommitOffset() is called effectively [within the committer](https://github.com/tmdiep/google-cloud-go/blob/6f7a727a628b79d580060aaa2ecdbc7ad723dae0/pubsublite/internal/wire/committer.go#L127) in a periodic task that allows batching of commits.

The Java client lib doesn't batch commits and it looks like the cursor server handles this batching/deduplication anyway. But I thought it would be a good idea, like we did for flow control tokens, as it added very little overhead in code complexity here in the client lib and minimizes the network traffic. WDYT?

Reply:
Btw, the batching behavior can be easily switched on/off with relatively little code changes.

Reply:
Sorry, I misunderstood your comment, which was absolutely correct. I've just inlined Pop() within CommitOffset(). Thanks!